lms/djangoapps/certificates/signals.py

```diff
@@ -1,29 +1,27 @@
-""" Signal handler for enabling self-generated certificates by default
-for self-paced courses.
+"""
+Signal handler for enabling/disabling self-generated certificates based on the course-pacing.
 """
 from celery.task import task
-from django.dispatch.dispatcher import receiver
+from django.dispatch import receiver
 from opaque_keys.edx.keys import CourseKey
 
 from certificates.models import CertificateGenerationCourseSetting
-from openedx.core.djangoapps.content.course_overviews.models import CourseOverview
-from xmodule.modulestore.django import SignalHandler
+from openedx.core.djangoapps.models.course_details import COURSE_PACING_CHANGE
 
 
-@receiver(SignalHandler.course_published)
-def _listen_for_course_publish(sender, course_key, **kwargs):  # pylint: disable=unused-argument
-    """ Catches the signal that a course has been published in Studio and
-    enable the self-generated certificates by default for self-paced
-    courses.
+@receiver(COURSE_PACING_CHANGE, dispatch_uid="course_pacing_changed")
+def _listen_for_course_publish(sender, course_key, course_self_paced, **kwargs):  # pylint: disable=unused-argument
+    """
+    Catches the signal that course pacing has changed and enable/disable
+    the self-generated certificates according to course-pacing.
     """
-    enable_self_generated_certs.delay(unicode(course_key))
+    toggle_self_generated_certs.delay(unicode(course_key), course_self_paced)
```

Comment:
we need to do this because data transferred between clients and workers needs to be serialized.
Here is the link:
http://docs.celeryproject.org/en/latest/userguide/calling.html#serializers